### ISSUE TITLE
setting multiple allowed-origins in CORS results in AWS APIG that defies W3C

### ISSUE THREAD

**Issue by jokeyrhyme** [serverless/serverless#2476]
# This is a Bug Report
- What went wrong?

TL;DR: Access-Control-Allow-Origin header contains multiple origins, which violates the W3C specification and makes Chrome cranky

There's a combination of things here...

The documentation and YAML structure imply that multiple allowed-origins for CORS settings are supported, because it is a list with plural name (origins), and not a string with a singular name (origin):

https://serverless.com/framework/docs/providers/aws/events/apigateway/#enabling-cors-for-your-endpoints

``` yaml
functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: user/create
          method: get
          cors:
            origins:
              - '*'
```

Serverless seems to configure a static HTTP response header for Access-Control-Allow-Origin, with a string value that is a comma-delimited version of the list that we specify in the YAML.

This is against the W3C CORS specification, and Chrome complains (in the console) when it sees `Access-Control-Allow-Origin: https://origin1.com,https://origin2.com`

https://www.w3.org/TR/cors/#resource-requests

> add a single Access-Control-Allow-Origin header, with either the value of the Origin header or the string "*" as value

The Access-Control-Allow-Origin header has to be a single origin that is the same as the incoming Origin header, or the wildcard "*". It is not allowed to be multiple origins. I would think that this design is to avoid information disclosure, to prevent a failed pre-flight check from telling malicious parties all about all the other origins that can try.

Unfortunately, AWS API Gateway itself only supports a single allowed-origin, probably to simplify implementation and improve the performance of the service:

https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html

> Use the console-provided value of '_*' as the Access-Control-Allow-Origin header value to allow access requests from all domains, or specify a named domain to all access requests from the specified domain

"a named domain" - singular
- What did you expect should have happened?

For example, say I allow 2 origins: https://origin1.com and https://origin2.com
- request from https://origin1.com = Access-Control-Allow-Origin: https://origin1.com
- request from https://origin2.com = Access-Control-Allow-Origin: https://origin2.com
- request from https://origin3.com = no CORS headers

Server-side web server frameworks like Hapi.js support configuring multiple allowed-origins, but only ever actually respond with the incoming request Origin header if there is a match:

https://github.com/hapijs/hapi/blob/v15.1.1/lib/cors.js#L85

Ideally, AWS APIG would support multiple headers to be configured, and perform this request-time match for us. But that doesn't seem available just now.

Alternatively, is it possible to achieve this sort of matching with an APIG response header template that Serverless defines?
- What was the config you used?

``` yaml
functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: user/create
          method: get
          cors:
            origins:
              - https://origin1.com
              - https://origin2.com
```
- What stacktrace or error message from your provider did you see?

n/a
## Additional Data
- **_Serverless Framework Version you're using**_: 1.0.0-rc.2
- **_Operating System**_: macOS Sierra

**Comment by jokeyrhyme** [comment]
@mymattcarroll

**Comment by pmuens** [comment]
Thank you very much for reporting @jokeyrhyme 👍 

That makes sense! I'm only concerned that this might break existing deployments when renaming `origins` to `origin` and only supporting a string config variable.

@brianneisler @eahefnawy do you guys have any idea how we can make this change in a non-breaking way?

Should we add the `origin` config parameter and keep `origins` until v2.0 is released? We can add a validation so that the user can only use `origin` xor `origins`. Any other ideas?

**Comment by brianneisler** [comment]
@pmuens that seems to be the simplest way to fix this in a backwards compatible way. 

> Alternatively, is it possible to achieve this sort of matching with an APIG response header template that Serverless defines?

This might be worth taking two minutes to look at... would be a great feature if we could get that to work.

**Comment by jokeyrhyme** [comment]
@pmuens my preference would be to:
-   short term: document the current behaviour with a warning
-   short term: detect when `origins` has more than 1 entry, and display a warning to the user
-   long term: keep `origins` and make it work

Even if we can't get this to work the way we want it to, I would discourage renaming the setting to the singular `origin`, for the compatibility reasons you point out.

**Comment by pmuens** [comment]
Hey everyone! I just PRed a WIP implementation --> https://github.com/serverless/serverless/pull/3692

Feel free to test it and provide feedback!

**Comment by alexcasalboni** [comment]
@pmuens #3692 is not fixing the problem completely, right?

What is the current recommended way to allow multiple origins? (e.g. https://mywebsite.com and http://local.mywebsite.com:8000)

The only solution I've found so far is implementing the OPTIONS method with a Lambda Function explicitly, checking the request origin against a list or trusted origins, and returning the given origin only if it's allowed (see [this answer on SO](https://stackoverflow.com/a/41708323/1875166)).

**Comment by jokeyrhyme** [comment]
@alexcasalboni this is my understanding, too, that to properly implement multiple origins we need a little dynamic behaviour. It's possible that this might be achieved with a template, but it might be necessary to wrap user code with a Serverless wrapper than intercepts OPTIONS requests and implements the proper CORS logic. Such a wrapper would open new doors for addressing shortcomings in AWS Lambda, but it may also fall quite outside the scope of what Serverless wants to offer.

**Comment by pmuens** [comment]
> @pmuens #3692 is not fixing the problem completely, right?

@alexcasalboni 🤔 yes, this only enables a way do define one origin.

Would be nice to discuss different implementation proposals here. Is there a CloudFormation / AWS supported way to do this?

Thanks @jokeyrhyme @alexcasalboni thanks for starting the proposal process 👍

**Comment by alexcasalboni** [comment]
@pmuens Exactly, AWS is not helping at all.

The only real solution I'm aware of requires implementing the OPTIONS method yourself. The framework could automatically create the Lambda Function that handles every OPTIONS request, given the list of trusted origins.

Drawbacks:

- You need N more trigger definitions (one for each path)
- You will pay for a Lambda execution for each OPTIONS request (even if the GET response is cached), and it's probably going to be a bit slower too


As a side note, this wouldn't solve the Custom Authorizer CORS problem (see [here](https://forums.aws.amazon.com/thread.jspa?messageID=728839)).

**Comment by HyperBrain** [comment]
@alexcasalboni It might be possible to implement OPTIONS endpoints as pure APIG implementation using templates and VTL. Serverless could create an OPTIONS template that contains an VTL if construct that would return the proper headers depending on the paths and methods of the request. If the allowed hosts are defined in the serverless.yaml file, the template could check the Origin header against that list and return the hostname in the allowed Origin header or return an appropriate Forbidden error.
With such an implementation OPTIONS would completely be handled within APIG. The endpoint itself should be a MOCK endpoint (just to eliminate the required lambda) and rewrite the integration response according to the configuration. The mock object itself can be empty.

**Comment by pmuens** [comment]
@alexcasalboni and @HyperBrain thanks for the feedback on this one.

#3692 was a fix so that the config parameters reflect what you'll get from AWS.

I've opened up https://github.com/serverless/serverless/issues/3745 so that we can discuss the implementation details in this dedicated issue.

**Comment by rodrigomf24** [comment]
If you are bashing your head against this mind-numbing problem, here is what worked for me, none of the "solutions" I encountered fixed the problem for me, but this one did:

https://blog.rob.uk.com/2019/01/26/Allowing-multiple-CORS-origins-with-AWS-API-Gateway/

Yes, it sucks to have a definition for every lambda, but it is the only solution that has successfully worked for me:

This is my implementation using typescript:

```
import {
  APIGatewayProxyCallback,
  APIGatewayProxyEventV2,
  Context,
} from 'aws-lambda';

const ALLOWED_ORIGINS_REGEX = [
  new RegExp(/https:\/\/(.*).somedomain.com/),
  new RegExp(/https:\/\/(.*).anotherdomain.com/),
]

export const getCorsOrigin = (incomingOrigin?: string) => {
  if (!incomingOrigin) {
    return null;
  }

  const isAllowedOrigin = ALLOWED_ORIGINS_REGEX.find((whitelistedOrigin) => {
    if (whitelistedOrigin instanceof RegExp) {
      return whitelistedOrigin.test(incomingOrigin);
    }

    return incomingOrigin === whitelistedOrigin;
  });

  if (incomingOrigin && isAllowedOrigin) {
    return incomingOrigin;
  }

  return null;
};

export const resolver = async (
  event: APIGatewayProxyEventV2,
  _context: Context,
  callback: APIGatewayProxyCallback
) => {
  let requestOrigin = event.headers["Origin"];

  if (!requestOrigin || !requestOrigin.length) {
    requestOrigin = event.headers["origin"];
  }

  const allowedOrigin = evaluateRequestOrigin(requestOrigin);

  const nextHeaders: Record<string, string | boolean | number> = {};

  if (allowedOrigin) {
    nextHeaders["Access-Control-Allow-Credentials"] = true;
    nextHeaders["Access-Control-Allow-Origin"] = allowedOrigin;
    nextHeaders["Access-Control-Allow-Headers"] =
      "Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token";
    nextHeaders["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS";
  }

  callback(null, {
    statusCode: 200,
    headers: nextHeaders,
    body: "",
  });
};

  ```
  
😮‍💨  I hope this helps anyone that stumbles into this thread...
